### ISSUE TITLE
EventBridge event setting arn with ImportValue causes TypeError

### ISSUE THREAD

**Issue by DaveLo** [serverless/serverless#7063]
# Bug Report

## Description

1. What did you do?
    - Trying to import a Custom event bus arn from another stack caused a type error.
1. What happened?
   - Failed with error
   - Retried using inline`!ImportValue` syntax also failed
   - Tried with a hardcoded arn and successfully deployed
1. What should've happened?
   - Stack deploy with event from imported bus
1. What's the content of your `serverless.yml` file?
```shell
Framework Core: 1.59.1
Plugin: 3.2.5
SDK: 2.2.1
Components Core: 1.1.2
Components CLI: 1.4.0
```

```yaml
service: Example

custom:
  env: ${file(.env.yml)} 

provider:
  name: aws
  profile: ${opt:aws-profile, self:custom.env.defaults.profile}
  stage: ${opt:stage, self:custom.env.defaults.stage}
  region: ${opt:region, self:custom.env.defaults.region}
  runtime: nodejs12.x

  functions:
    testFunc:
      handler: src/this/is/a/test.handler
      events:
        - eventBridge:
            eventBus: 
              Fn::ImportValue: OtherStackBusArn
            # also tried - 
            #  eventBus: !ImportValue OtherStackBusArn
            pattern:
              source:
                - event.test
```
1. What's the output you get when you use the `SLS_DEBUG=*` environment variable (e.g. `SLS_DEBUG=* serverless deploy`)

```shell
TypeError: EventBus.startsWith is not a function
      at /######/.nvm/versions/node/v12.13.1/lib/node_modules/serverless/lib/plugins/aws/package/compile/events/eventBridge/index.js:121:30
      at Array.forEach (<anonymous>)
      at /######/.nvm/versions/node/v12.13.1/lib/node_modules/serverless/lib/plugins/aws/package/compile/events/eventBridge/index.js:29:28
      at Array.forEach (<anonymous>)
      at AwsCompileEventBridgeEvents.compileEventBridgeEvents (/######/.nvm/versions/node/v12.13.1/lib/node_modules/serverless/lib/plugins/aws/package/compile/events/eventBridge/index.js:23:31)
      at /######/.nvm/versions/node/v12.13.1/lib/node_modules/serverless/lib/classes/PluginManager.js:489:55
  From previous event:
      at PluginManager.invoke (/######/.nvm/versions/node/v12.13.1/lib/node_modules/serverless/lib/classes/PluginManager.js:489:22)
      at PluginManager.spawn (/######/.nvm/versions/node/v12.13.1/lib/node_modules/serverless/lib/classes/PluginManager.js:509:17)
      at Deploy.<anonymous> (/######/.nvm/versions/node/v12.13.1/lib/node_modules/serverless/lib/plugins/deploy/deploy.js:115:50)
  From previous event:
      at Object.before:deploy:deploy [as hook] (/######/.nvm/versions/node/v12.13.1/lib/node_modules/serverless/lib/plugins/deploy/deploy.js:100:30)
      at /######/.nvm/versions/node/v12.13.1/lib/node_modules/serverless/lib/classes/PluginManager.js:489:55
  From previous event:
      at PluginManager.invoke (/######/.nvm/versions/node/v12.13.1/lib/node_modules/serverless/lib/classes/PluginManager.js:489:22)
      at /######/.nvm/versions/node/v12.13.1/lib/node_modules/serverless/lib/classes/PluginManager.js:524:24
  From previous event:
      at PluginManager.run (/######/.nvm/versions/node/v12.13.1/lib/node_modules/serverless/lib/classes/PluginManager.js:524:8)
      at /######/.nvm/versions/node/v12.13.1/lib/node_modules/serverless/lib/Serverless.js:115:33
      at processImmediate (internal/timers.js:439:21)
      at process.topLevelDomainCallback (domain.js:130:23)
  From previous event:
      at Serverless.run (/######/.nvm/versions/node/v12.13.1/lib/node_modules/serverless/lib/Serverless.js:102:74)
      at /######/.nvm/versions/node/v12.13.1/lib/node_modules/serverless/bin/serverless.js:72:30
      at /######/.nvm/versions/node/v12.13.1/lib/node_modules/serverless/node_modules/graceful-fs/graceful-fs.js:111:16
      at /######/.nvm/versions/node/v12.13.1/lib/node_modules/serverless/node_modules/graceful-fs/graceful-fs.js:45:10
      at FSReqCallback.oncomplete (fs.js:146:23)
  From previous event:
      at /######/.nvm/versions/node/v12.13.1/lib/node_modules/serverless/bin/serverless.js:72:8
      at processImmediate (internal/timers.js:439:21)
      at process.topLevelDomainCallback (domain.js:130:23)
  From previous event:
      at Object.<anonymous> (/######/.nvm/versions/node/v12.13.1/lib/node_modules/serverless/bin/serverless.js:61:4)
      at Module._compile (internal/modules/cjs/loader.js:959:30)
      at Object.Module._extensions..js (internal/modules/cjs/loader.js:995:10)
      at Module.load (internal/modules/cjs/loader.js:815:32)
      at Function.Module._load (internal/modules/cjs/loader.js:727:14)
      at Function.Module.runMain (internal/modules/cjs/loader.js:1047:10)
      at internal/main/run_main_module.js:17:11
```

Similar or dependent issues:

- [This issue](https://forum.serverless.com/t/amazon-eventbridge-partner-event-sources-error/9011) from the forum about partner events seems at least tangentially related

**Comment by DaveLo** [comment]
Hey, I thought I might take a crack at this. I looked at how it's handled in SQS events, but I don't think it tracks exactly.

From looking at the code it looks like we always want the event bus name not the arn so that it can be used in constructing the rule arn. maybe it's easier to import the name by exporting  the external Ref (name for EventBridge)?
Then you could look for Fn::ImportValue (`(typeof EventBus !== 'string' && 'Fn::ImportValue' in EventBus)` from sqs event) and workflow from there?  

```javascript
ruleResource = {
  'Fn::Join': [
    ':',
    [
      'arn:aws:events',
      { Ref: 'AWS::Region' },
      { Ref: 'AWS::AccountId' },
      { 
        'Fn::Join': [
          "/",
          [
            'rule',
            {'Fn::ImportValue':  EventBus['Fn::ImportValue' },
            RuleName
          ]
        ]
      }
    ],
  ],
};
```

**Comment by DaveLo** [comment]
I may need a little direction (paging @medikoo or @pmuens )

```javascript
if (typeof EventBus === 'string' && EventBus.startsWith('arn')) {
  eventBusName = EventBus.slice(EventBus.indexOf('/') + 1);
}
eventBusResource = {
  'Fn::Join': [
    ':',
    [
      'arn',
      { Ref: 'AWS::Partition' },
      'events',
      { Ref: 'AWS::Region' },
      { Ref: 'AWS::AccountId' },
      {
        'Fn::Join': ['/', ['event-bus', eventBusName]],
      },
    ],
  ],
};
ruleResource = {
  'Fn::Join': [
    ':',
    [
      'arn',
      { Ref: 'AWS::Partition' },
      'events',
      { Ref: 'AWS::Region' },
      { Ref: 'AWS::AccountId' },
      {
        'Fn::Join': ['/', ['rule', eventBusName, RuleName]],
      },
    ],
  ],
};
```
seems to fix the immediate issue, but exporting the Ref rather than the full Arn it seems to not realize the eventbus already exists. 

> An error occurred: TestFuncCustomEventBridge1 - Failed to create resource. Event bus TestBus already exists. 

Best guess is that in some part of the plumbing the imported values are being resolved and since the Ref isn't an arn `_.merge(compiledCloudFormationTemplate.Resources, customEventBridge);` maybe is not resolving correctly.

If there is a way to split an imported arn to get a part of it then I think I could just fold that in as additional conditional, but I'm not familiar enough with the code to know if that exists or how imports are handled  ( tougher first issue than anticipated ;p ). Otherwise I think maybe an optional property for importedArns could work, but feels clunky.

**Comment by medikoo** [comment]
Thanks @DaveLo for looking into that.

Firstly, this issue comes from fact, that CloudFormation instructions are not normally supported on serverless configuration level. Technically they can be supported only if given setting value is propagated 1:1 to CloudFormation template property (where given instruction is supported), but it requires further effort to implement such support.

Ideally Framework should fail with meaningful error on such attempts, but it doesn't (hence ugly errors as `EventBus.startsWith is not a function`). We plan to address that at #6562 

Now, concerning your proposal (on bringing support for CF instruction format to `eventBus` property):

> but exporting the Ref rather than the full Arn it seems to not realize the eventbus already exists.

Currently the framework assumes event bus exists, only if value passed to `eventBus` resolves to string that starts with `arn`. See, this setting is propagated in 1:1 form to custom resource: 

https://github.com/serverless/serverless/blob/2938a95b8711d3bfab1e665318f46f66db91d6a4/lib/plugins/aws/package/compile/events/eventBridge/index.js#L94

and then read here:

https://github.com/serverless/serverless/blob/2938a95b8711d3bfab1e665318f46f66db91d6a4/lib/plugins/aws/customResources/resources/eventBridge/lib/eventBridge.js#L10

**Comment by DaveLo** [comment]
That makes sense @medikoo , adding a human readable error there would be a reasonable fix to limit confusion. And I can see how not having the import be 1:1 could be an issue given how the resources are built.

For my use case it's basically always going to be an imported bus (communicating between services that have no direct relations) and in a aws service designed for exactly that it feels counterintuitive that I would be copying and pasting an arn into every service subscribed to an eventBus creating an explicit link between code bases rather than a referential one. 

I feel like that implementation would be beyond my current understanding of the framework, but just to throw out a workflow maybe adding an optional eventBus object pattern like:
```yaml
- eventBridge:
    eventBus:
      arn: !ImportValue testBusArn
      name: !ImportValue testBusName
    pattern: 
      account: !Ref AWS::AccountId
```
to keep explicit 1:1 on import use? The string for simple use, object for more complex cases pattern is pretty common in most of the framework so it wouldn't feel out of place IMO

**Comment by medikoo** [comment]
@DaveLo I've looked into that more deeply, and to me it appears that there's inconsistency in our config design.

e.g. for S3 buckets or cognito user pools, resources are referenced by name, and via additional `existing: true` setting we eventually mark that given resource already exists, so there should not be creation attempt.

While in case of event bridge, we combine `existing: true` into name setting (if it's an ARN, we imply `existing: true`, and if not, then we create).

I believe right way to approach it, is to bring same `existing: true` setting support to event bridge configuration (and ensure that that CF instructions are supported for name), that way it'll work for you as expected.
(Unfortunately to not break things in v1, we still need to support ARN resolution).

@pmuens What do you think about that?

**Comment by Mario-Rod** [comment]
I have the same problem would be solved with the mentioned by @medikoo, do you plan to incorporate this parameter "existing: true"?

**Comment by DaveLo** [comment]
I gave this another try today, it seems like I'm missing something important. 
```javascript
                if (!busExists) {
                  eventBusResource = {
                    'Fn::Join': [
                      ':',
                      [
                        'arn',
                        { Ref: 'AWS::Partition' },
                        'events',
                        { Ref: 'AWS::Region' },
                        { Ref: 'AWS::AccountId' },
                        {
                          'Fn::Join': ['/', ['event-bus', eventBusName]],
                        },
                      ],
                    ],
                  };
                  iamRoleStatements.push({
                    Effect: 'Allow',
                    Resource: eventBusResource,
                    Action: ['events:CreateEventBus', 'events:DeleteEventBus'],
                  });
                }
```
Gets through and stops the error, but then I get a permissions error for what might be a custom cloudformation thing that the framework is doing `not authorized to perform: events:CreateEventBus`. I'm guessing it's custom because CreateEventBus isn't a cloudformation directive[ `events:EventBus`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-events-eventbus.html) is.

I know there is another ticket about moving over to CF (#7148 ), but am not certain whether that is as easy as just changing it in one file.

**Comment by medikoo** [comment]
> not authorized to perform: events:CreateEventBus

If you mean that this crash happens in custom resource lambda (during deployment), then it means that setup of access rules  which this lambda needs, needs to be updated here: 

https://github.com/serverless/serverless/blob/fbf99fa2abf9ce3bc13fc4a6c8439a650d3eaa4e/lib/plugins/aws/package/compile/events/eventBridge/index.js#L152-L168

**Comment by DaveLo** [comment]
> If you mean that this crash happens in custom resource lambda (during deployment), then it means that setup of access rules which this lambda needs, needs to be updated here:

That is the section I'm altering for an existing bus. Having played with moving things around a bit to see what happens, if I move the Iam statement push for `CreateEventBus` out of the conditional I've added for existing I get a resource error since it's trying to apply a policy to an undefined var (`eventBusResource`). 

I initially moved it into the same conditional statement, for least permissions principle since I don't need the ability to create an event bus if one exists. I think this is where the custom resource lambda part comes in, it looks to be one handler with a conditional chain for what it is supposed to do (create,update,delete). My guess on initial look is that it needs all of those permissions because it's one function rather than one handler per execute type.

https://github.com/serverless/serverless/blob/59634e7d926e1c7d5173e34e07e14bfdebf49dc9/lib/plugins/aws/customResources/resources/eventBridge/handler.js#L15-L23

I could split this into separate EventBus and EventRule custom resources, but if the framework is going to update implementation to use CF it might just be worth it to wait and proceed with the 'existing' feature after?

**Comment by medikoo** [comment]
@DaveLo I would definitely keep one lambda, and in all cases provide all permissions it needs for all eventual cases. It doesn't seem to me it may imply any security issue.

**Comment by DaveLo** [comment]
Looking through the custom resource function code and:

https://github.com/serverless/serverless/blob/59634e7d926e1c7d5173e34e07e14bfdebf49dc9/lib/plugins/aws/customResources/resources/eventBridge/lib/utils.js#L3-L8

I double checked the AWS SDK  to be sure, but I can't import values inside a lambda. So I think this one is going to need to wait on updating the EventBridge code to use CloudFormation. I'm not sure my knowledge level of the framework internals is up to that tbh.

**Comment by medikoo** [comment]
> I double checked the AWS SDK to be sure, but I can't import values inside a lambda.

You mean values from given CloudFormation stack (which we deploy), or some outputs as exported by other stack?

**Comment by DaveLo** [comment]
> You mean values from given CloudFormation stack (which we deploy), or some outputs as exported by other stack?

The second one. 

The custom resource lambda  is using the code in my last comment to get the event bus name. If I were to pass  `eventBus = { "Fn::ImportValue": "testName" }` to that lambda, I wouldn't be able to get the value even if I updated the conditional to handle objects because the AWS-SDK CloudFormation class doesn't have a get export method. 

https://github.com/serverless/serverless/blob/59634e7d926e1c7d5173e34e07e14bfdebf49dc9/lib/plugins/aws/package/compile/events/eventBridge/index.js#L82-L103

Where it looks like the custom resource is added to the stack it is passing in the EventBus variable, which in the case we are trying to solve for would be an importValue. 

Unless I'm missing something about the framework resolving AWS CF intrinsic functions locally before deploying a custom resource?

**Comment by medikoo** [comment]
>  If I were to pass eventBus = { "Fn::ImportValue": "testName" } to that lambda, I wouldn't be able

You may use AWS SDK in a lambda to retrieve that import. See here, how it can be done: https://github.com/serverless/serverless/blob/482f0356af77185594de949d4969e59da39054ed/lib/plugins/aws/package/compile/events/apiGateway/lib/hack/updateStage.js#L90-L97

**Comment by danrivett** [comment]
I ran into the same problems today: Both trying to import the EventBus name from a named export as well as subscribing a lambda to an existing EventBus instance rather than creating another one.

Not sure if this issue is going to be resolved or if the workaround below is expected to be used medium term?

To summarize the above conversation as I understand it, the workaround for both these issues is to specify the EventBus arn and not the name, and to do so using a string the following format:

```
arn:aws:events:#{AWS::Region}:#{AWS::AccountId}:event-bus/<eventbus-name>
```

**Comment by medikoo** [comment]
@danrivett we're open for PR that adds support for a CloudFormation instruction

**Comment by stuartforrest-infinity** [comment]
@DaveLo It sounds like you have done a bunch of work begin to facilitate the use of `CF funcs` to define the `eventBus` against a function. Do you have a branch you can push up I could work off of? I'm looking at implementing support for this.

**Comment by DaveLo** [comment]
> @DaveLo It sounds like you have done a bunch of work begin to facilitate the use of `CF funcs` to define the `eventBus` against a function. Do you have a branch you can push up I could work off of? I'm looking at implementing support for this.

I did some of the initial digging into why it was happening, but I wasn't able to successfully implement a fix (I think because I'm not well versed in how the create custom resource stuff works). I have a very stale branch in my profile for [`fix/eventbridge-import-bus`](https://github.com/DaveLo/serverless/tree/fix/eventbridge-import-bus) where I was starting to attempt adding the existing flag, but have been too buried in other dayjob work to get back to :-/ 

I feel like the better avenue long term would be to move to using CloudFormation since all of EventBridge bus and rule definition is now possible there. That would probably be a much bigger task, but no custom resources necessary.

**Comment by stuartforrest-infinity** [comment]
@DaveLo Having had a further dig using `CloudFormation` does look like the right thing to do. @medikoo What are your thoughts on this?

**Comment by stuartforrest-infinity** [comment]
@medikoo Is there a precedent for switching from custom resource to CF and handling deletion/importing of custom resources?

**Comment by medikoo** [comment]
> @medikoo Is there a precedent for switching from custom resource to CF and handling deletion/importing of custom resources?

Good question! No, I believe there's not.

If we can't do it without breaking deployment, then I believe we need to implement this feature with graceful adaptation in mind, as follows:
- Add support to configure event bridge via plain CF, but as opt-in (to use it, developer needs to turn it on in config)
- Show [deprecation log](https://github.com/serverless/serverless/blob/master/CONTRIBUTING.md#configuring-deprecations) when someone relies on custom resource event bridge generation
- With next major release remove custom resource based setup, and leave just plain CF way

**Comment by stuartforrest-infinity** [comment]
Attempting to import the resources in the the stack within the framework may not be possible - https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/resource-import-supported-resources.html - according to these docs `EventBus` and `lambda:Permission` are not resource types that can be imported. 

Even if this is out of date and import is possible, I think we would need to execute a `ChangeSet` operation against the stack with the old stack definition before attempting to deploy any changes to the stack. I think this might be more complex an operation that the framework wants to be responsible for but, having said that, what we do here may set the precedent for future `Custom` to `CF` migrations.

I think it will introduce complexity that may be undesirable - what are the thoughts from others?

The alternative is to provide instructions for users to manually import the resources and handle this in the `delete` handle for the custom resource - probably just do nothing as the the `DeletionPolicy` is almost certainly going to be set to `Retain`.

**Comment by medikoo** [comment]
@stuartforrest-infinity As I mentioned I think we should just introduce another opt-in option, which will default with next major version.

So for users that have already `eventBridge` events deployed, the only way to (optionally) migrate, would be to remove their stacks and recreate them with new settings.

I don't think we should put an effort in attempting to seamlessly move already created resources. That doesn't seem possible (at least not easily)

**Comment by AntonyPritchett** [comment]
@medikoo I'm having issues with exactly this. Could a workaround be to just create the rule in CF and refer to the serverless function as a Target?

**Comment by medikoo** [comment]
>  I think guidance on importing the resources can be provided and potentially a helper script to output the required CloudFormation 

@stuartforrest-infinity if it's the case, then definitely we can include that, and all the instructions can be put into documentation of _deprecation_ (we document each one here: https://www.serverless.com/framework/docs/deprecations/)

**Comment by XM37** [comment]
What is the current situation regarding this case? Is an update planned?

**Comment by medikoo** [comment]
@XM37 We' re open for PR that'll bring support for this on CloudFormation level, afaik no one works on it. Do you want to help with that?

**Comment by stuartforrest-infinity** [comment]
There is work on this here - https://github.com/stuartforrest-infinity/serverless/tree/handle-cf-funcs-eventbridge

The branch name is misleading, the work is to use native Cloud Formation for the resources. I was in the middle of updating/writing tests if I remember rightly...

**Comment by stuartforrest-infinity** [comment]
@medikoo I have completed work on the branch linked to above to deploy EventBridge resources using CloudFormation. I have no doubt there will be changes needed as it is not a trivial change and I've had to do a lot learning about how the framework is operation to complete it. Are you happy for me to create a PR off of it and we can go from there?

**Comment by medikoo** [comment]
> Are you happy for me to create a PR off of it and we can go from there?

@stuartforrest-infinity definitely, please open PR, we look forward that change!

**Comment by cjuega** [comment]
I'm experiencing issues with this implementation in **serverless@v2.28.0**. Here is my setup:

```YAML
provider:
  ...
  eventBridge:
    useCloudFormation: true
```

I have the following resource defined:

```YAML
Resources:
  DomainEventBus:
    Type: AWS::Events::EventBus
    Properties:
      Name: ${env:PROJECT}-${env:STAGE}-events

Outputs:
  DomainEventBusArn:
    Value: !GetAtt DomainEventBus.Arn
    Export:
      Name: ${env:PROJECT}-${self:provider.stage}-DomainEventBusArn
```

Then, in my lambda definitions, I use `!GetAtt DomainEventBus.Arn` or `!ImportValue DomainEventBusArn` to indicate which eventbus triggers a lambda:

```YAML
  events:
    - eventBridge:
        eventBus: !GetAtt DomainEventBus.Arn
```

Deployment works fine. It seems everything is correct. However, my lambdas are not triggered when events are sent to the bus. After some troubleshooting I have found lambdas' **Resource-based policy** is wrong. In particular, `AWS:SourceArn` condition has a wrong ARN. It seems eventbus' name is not extracted from the ARN in order to build rule's ARN.

The quick workaround is to use `!Ref DomainEventBus` to use eventbus' name instead of its ARN.

**Comment by pgrzesik** [comment]
Hey @cjuega - how did you use `arn`s previously? Via `Fn::Import` as well or passing it explicitly?

**Comment by cjuega** [comment]
@pgrzesik, do you mean before using the new `useCloudFormation: true` feature? I used to pass it explicitly in an envvar. Using `!ImportValue` thrown an `EventBus.startsWith is not a function` error as described in the bug report.

**Comment by capgadsx** [comment]
I'm facing the same problem of @cjuega.

Before of using `useCloudFormation: true` the generated policy was like this:

```json
{
    "Version": "2012-10-17",
    "Id": "default",
    "Statement": [
        {
            "Sid": "[long name]",
            "Effect": "Allow",
            "Principal": {
                "Service": "events.amazonaws.com"
            },
            "Action": "lambda:InvokeFunction",
            "Resource": "arn:aws:lambda:[zone]:[account_id]:function:[function_name]",
            "Condition": {
                "ArnLike": {
                    "AWS:SourceArn": "arn:aws:events:[zone]:[account_id]:rule/[bus_name]/[rule_name]"
                }
            }
        }
    ]
}
```

Now if we use:

```yaml
FunctionName:
  handler: lambda_function.handler
  events:
    - eventBridge:
        eventBus: !GetAtt InputEventBus.Arn
        pattern:
          source:
            - "pattern"
```

The generated policy is the following:

```json
{
    "Version": "2012-10-17",
    "Id": "default",
    "Statement": [
        {
            "Sid": "[long name]",
            "Effect": "Allow",
            "Principal": {
                "Service": "events.amazonaws.com"
            },
            "Action": "lambda:InvokeFunction",
            "Resource": "arn:aws:lambda:[zone]:[account_id]:function:[function_name]",
            "Condition": {
                "ArnLike": {
                    "AWS:SourceArn": "arn:aws:events:[zone]:[account_id]:rule/[full_bus_arn]/[rule_name]"
                }
            }
        }
    ]
}
```

@cjuega workarround solves the problem. The generated CF template goes from here.

```json
"SourceArn": {
  "Fn::Join": [
    ":",
    [
      "arn",
      {
        "Ref": "AWS::Partition"
      },
      "events",
      {
        "Ref": "AWS::Region"
      },
      {
        "Ref": "AWS::AccountId"
      },
      {
        "Fn::Join": [
          "/",
          [
            "rule",
            {
              "Fn::GetAtt": [
                "InputEventBus",
                "Arn"
              ]
            },
            "[rule_name]"
          ]
        ]
      }
    ]
  ]
}
```

To this.

```json
"SourceArn": {
  "Fn::Join": [
    ":",
    [
      "arn",
      {
        "Ref": "AWS::Partition"
      },
      "events",
      {
        "Ref": "AWS::Region"
      },
      {
        "Ref": "AWS::AccountId"
      },
      {
        "Fn::Join": [
          "/",
          [
            "rule",
            {
              "Ref": "InputEventBus"
            },
            "[rule_name]"
          ]
        ]
      }
    ]
  ]
}
```

**Comment by pgrzesik** [comment]
Thank you both @cjuega and @capgadsx - I believe I know what the issue is here. In order to generate proper `AWS:SourceArn` we need to have the ability to extract `name` of the event bus. It was possible to do when using explicit `arn` passed as a string, however, when passing arn with CF intrinsic command, it's impossible to reason about `name`, which results in invalid `SourceArn` being generated. The proper approach in that situation would be to pass imported name as suggested in workaround by @cjuega. I believe we should clarify that in the documentation.

**Comment by cjuega** [comment]
Glad to help!

Just to recap, the workaround is not to use an imported value but passing the eventbus' name instead of the ARN.

This virtually removes the ability of using eventbus' ARN (it will only work when it's hardcoded). I believe that's fine as we can still use eventbus' name. But, yes, it should be clarified in the documentation.

Summing up:

```YAML
- eventBridge:
    eventBus: arn:aws:events:us-east-1:12345:event-bus/custom-private-events   # this will work! :)


- eventBridge:
    eventBus: custom-private-events   # this will work! :)


- eventBridge:
    eventBus: !Ref EventBus   # this will work! :)


- eventBridge:
    eventBus: !GetAtt EventBus.Name   # this will work! :)


- eventBridge:
    eventBus: !ImportValue EventBusName   # this will work as long as the Output refers to eventbus' name! :|


- eventBridge:
    eventBus: !GetAtt EventBus.Arn   # this WON'T work! :(
```

**Comment by tiagoramalho** [comment]
Hey @cjuega, @pgrzesik 

I wanted to use the first option, but I wanted to set the account dynamically,
Do you know if this should be supported? 

`eventBus: !Sub arn:aws:events:us-east-1:${AWS::AccountId}:event-bus/custom-private-events `

**Comment by pgrzesik** [comment]
Hey @tiagoramalho - looking at the schema definition it won't be accepted as `Sub` is not allowed, but I think it should safely extend it to support it as I think `Sub` will work just fine. Would you be interested in submitting a PR for that?

**Comment by 596050** [comment]
Hi, I couldn't find it clearly documented anywhere but what seemed to resolve the issue for me is:

For the Variables which need to be exposed/exported in outputs, they must have an "Export" property with a name sub-property:
 
```json
resources: {
    Resources: resources["Resources"],
    Outputs: {
     // For eventbus
      EventBusName: {
        Export: {
          Name: "${self:service}-${self:provider.stage}-UNIQUE_EVENTBUS_NAME",
        },
        Value: {
          Ref: "UNIQUE_EVENTBUS_NAME",
        },
      },
     // For something like sqs, or anything else, would be the same
      IDVerifyQueueARN: {
        Export: {
          Name: "${self:service}-${self:provider.stage}-UNIQUE_SQS_NAME",
        },
        Value: { "Fn::GetAtt": ["UNIQUE_SQS_NAME", "Arn"] },
      }
    },
  },
```

Once this is deployed you can check if the exports are present by running in the terminal:

```bash
aws cloudformation list-exports
```

Then there should be a Name property in a list:

```json
 {
            "ExportingStackId": "***",
            "Name": "${self:service}-${self:provider.stage}-UNIQUE_EVENTBUS_NAME", <-- same as given above (but will be populated with your service and stage)
            "Value": "***"
}
```

And then if the above is successful, you ca reference it with "Fn::ImportValue" like so, e.g.:

```json
Resource: {
        "Fn::ImportValue":
          "${self:service}-${self:provider.stage}-UNIQUE_EVENTBUS_NAME",  <-- same as given above (but will be populated with your service and stage)
}
```
